languagetool-language-modules/ar/src/main/java/org/languagetool/tagging/ar/ArabicTagger.java

```diff
@@ -0,0 +1,207 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+package org.languagetool.tagging.ar;
+
+import morfologik.stemming.DictionaryLookup;
+import morfologik.stemming.IStemmer;
+import org.jetbrains.annotations.Nullable;
+import org.languagetool.AnalyzedToken;
+import org.languagetool.AnalyzedTokenReadings;
+import org.languagetool.tagging.BaseTagger;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.lang.StringBuilder;
+import java.util.Locale;
+
+public class ArabicTagger extends BaseTagger {
+
+  public ArabicTagger() {
+    super("/ar/arabic.dict", new Locale("ar"));
+  }
+
+  @Override
+  public String getManualAdditionsFileName() {
+    return "/ar/added.txt";
+  }
+
+  /* Add the flag to an encoded tag */
+  public String addTag(String postag, String flag) {
+    StringBuilder tmp = new StringBuilder(postag);
+    if (flag == "W") {
+      tmp.setCharAt(postag.length() - 3, 'W');
+    } else if (flag == "K") {
+      tmp.setCharAt(postag.length() - 2, 'K');
+    } else if (flag == "L") {
+      tmp.setCharAt(postag.length() - 2, 'L');
+    }
+    return tmp.toString();
+
+  }
+
+  /* test if flag exists */
+  public boolean hasFlag(String postag, String flag) {
+    StringBuilder tmp = new StringBuilder(postag);
+    if (flag == "has_pronoun") {
+      return postag.endsWith("H");
+    } else if (flag == "has_jar") {
+      return postag.contains("B");
+    }
+
+    return false;
+
+  }
+
+  @Override
+  public List<AnalyzedTokenReadings> tag(final List<String> sentenceTokens) {
+
+    final List<AnalyzedTokenReadings> tokenReadings = new ArrayList<>();
+    final IStemmer dictLookup = new DictionaryLookup(getDictionary());
+    int pos = 0;
+    for (String word : sentenceTokens) {
+      final List<AnalyzedToken> l = new ArrayList<>();
+      List<AnalyzedToken> taggerTokens = asAnalyzedTokenListForTaggedWords(word, getWordTagger().tag(word));
+      addTokens(taggerTokens, l);
+      // additional tagging with prefixes
+      if (l.isEmpty()) {
+        addTokens(additionalTags(word, dictLookup), l);
+      }
+      if (l.isEmpty()) {
+        l.add(new AnalyzedToken(word, null, null));
+      }
+      tokenReadings.add(new AnalyzedTokenReadings(l, pos));
+      pos += word.length();
+    }
+    return tokenReadings;
+  }
+
+
+  @Nullable
+  protected List<AnalyzedToken> additionalTags(String word, IStemmer stemmer) {
+    List<AnalyzedToken> additionalTaggedTokens = new ArrayList<>();
+
+    //Any well-formed word started by conjuction letters like Waw and Feh is the same word without  the conjuction + a conjuction tag
```

Comment:
* Let's put a space after // to be consistent with other comments.
* conjuction → conjunction (twice in the sentance)
* Put only 1 space instead of 2 between words "without  the"